### ISSUE TITLE
Build stage should allow exec or null resources

### ISSUE THREAD

**Issue by spuder** [hashicorp/waypoint#617]
At my organization we have multiple deployment methods (Capistrano, Nomad, Kubernetes, Salt). 
The Capistrano deploys are the oldest and are effectively just using `scp` to copy code to the servers. For these applications there is no 'build' step per-say. 

According to the documentation, waypoint supports the following plugins for the `build stage`

- Docker Build
- Docker Pull Build
- Cloud Native Buildpacks
- Custom (Write your own)

https://www.waypointproject.io/docs/lifecycle/build

Request: 

Allow for legacy apps to be converted to waypoint by providing a `null` or `exec` driver for the `build` stage, without requiring the end user to write their own plugin.

**Comment by mitchellh** [comment]
Agreed. The main issue with this is making the output types match as you'd want for the deployment step. That is why I didn't do this for 0.1... but it something we want to support.

**Comment by krantzinator** [comment]
Can you help us understand what you are deploying in this workflow? If you were to use an `exec` deploy plugin, what information would you want to pass to that plugin?

**Comment by spuder** [comment]
An example for an `exec` step for Capistrano compatible deploy might be running an arbitrary script or rake task. 

```
./bin/setup
```

```
bundle exec rake:setup
```

**Comment by evanphx** [comment]
@spuder In that case, what would information would the build exec plugin output that would be used by the deploy plugin? A local directory path? A docker image id? It's this flow of information that is unclear to us presently.

**Comment by RobbieMcKinstry** [comment]
Following up on this for my use case, another framing of the `null` builder would be to make having a default `builder` optional. One use case would be using Waypoint to manage deployments of off-the-shelf Helm charts.

For example:

```hcl
project = "consul"
app "consul" {
    build {
        use "docker" {}
    } 

    deploy {
        use "helm" {
            name = "consul"
            chart = "consul"
            repository = "https://helm.releases.hashicorp.com"
            namespace = "consul"
        }
    }
}
```

This is what I have currently, which uses an empty Dockerfile just to get past the build step. One benefit of using Waypoint for off-the-shelf Helm charts is that you get GitOps out of the box, as Waypoint can observe your VCS provider like ArgoCD or similar tools.

It would be a better developer experience to omit the `build{}` stanza. This would decrease the runtime of `waypoint up` and remove the necessary of having an empty Dockerfile for `deploy` steps that don't require an artifact input.

**Comment by evanphx** [comment]
Hi @RobbieMcKinstry,

That's good insight! We're thinking a lot about making the model of Waypoint more flexible and this is a good example of that.

**Comment by arbourd** [comment]
We are using Waypoint almost like [Flux](https://fluxcd.io/). I think Waypoint is the perfect platform for getting things onto Nomad and have no need for the build step.

**Comment by KieranP** [comment]
Same here. In the interim, I've just used a dummy docker pull.

```
  build {
    use "docker-pull" {
      image = "hello-world"
      tag = "latest"
    }
  }
```

Not ideal, because it adds a few seconds to the deploy, but unavoidable until Waypoint accepts an app without a build step
